cpp/src/dictionary/detail/concatenate.cu

```diff
@@ -0,0 +1,268 @@
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/dictionary/detail/concatenate.hpp>
+#include <cudf/dictionary/dictionary_column_view.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/transform_scan.h>
+#include <algorithm>
+#include <rmm/device_uvector.hpp>
+#include <vector>
+
+namespace cudf {
+namespace dictionary {
+namespace detail {
+namespace {
+
+/**
+ * @brief Keys and indices offsets values.
+ *
+ * The first value is the keys offsets and the second values is the indices offsets.
+ * These are offsets to the beginning of each input column after concatenating.
+ */
+using offsets_pair = thrust::pair<size_type, size_type>;
+
+/**
+ * @brief Utility for calculating the offsets for the concatenated child columns
+ *        of the output dictionary column.
+ */
+struct compute_children_offsets_fn {
+  /**
+   * @brief Create the utility functor.
+   *
+   * The columns vector is converted into vector of column_view pointers so they
+   * can be used in thrust::transform_exclusive_scan without causing the
+   * compiler warning/error: "host/device function calling host function".
+   *
+   * @param columns The input dictionary columns.
+   */
+  compute_children_offsets_fn(std::vector<column_view> const& columns)
+    : columns_ptrs{columns.size()}
+  {
+    std::transform(
+      columns.begin(), columns.end(), columns_ptrs.begin(), [](auto& cv) { return &cv; });
+  }
+
+  /**
+   * @brief Return the first keys().type of the dictionary columns.
+   */
+  data_type get_keys_type()
+  {
+    dictionary_column_view dict_view(**std::find_if(
+      columns_ptrs.begin(), columns_ptrs.end(), [](auto pcv) { return pcv->size() > 0; }));
+    return dict_view.keys().type();
+  }
+
+  /**
+   * @brief Create the offsets pair for the concatenated columns.
+   *
+   * Both vectors have the length of the number of input columns.
+   * The sizes of each child (keys and indices) of the individual columns
+   * are used to create the offsets.
+   *
+   * @param stream Stream used for allocating the output rmm::device_uvector.
+   * @return Vector of offsets_pair objects for keys and indices.
+   */
+  rmm::device_uvector<offsets_pair> create_children_offsets(cudaStream_t stream)
+  {
+    std::vector<offsets_pair> offsets(columns_ptrs.size());
+    thrust::transform_exclusive_scan(
+      thrust::host,
+      columns_ptrs.begin(),
+      columns_ptrs.end(),
+      offsets.begin(),
+      [](auto pcv) {
+        dictionary_column_view view(*pcv);
+        return offsets_pair{view.keys_size(), view.size()};
+      },
+      offsets_pair{0, 0},
+      [](auto lhs, auto rhs) {
+        return offsets_pair{lhs.first + rhs.first, lhs.second + rhs.second};
+      });
+    auto d_offsets = rmm::device_uvector<offsets_pair>(offsets.size(), stream);
+    CUDA_TRY(cudaMemcpyAsync(d_offsets.data(),
```

Comment:
Careful here. This copy may not occur until after the function returns, which means `offsets` would be destroyed and no longer valid to copy from. To be safe, this function really needs to sync the stream.